compliance/collection/main.go

```diff
@@ -182,17 +217,106 @@ func manageNodeScanLoop(ctx context.Context, rescanInterval time.Duration, scann
 	msg := &sensor.MsgFromCompliance{
 		Node: nodeName,
-		Msg:  &sensor.MsgFromCompliance_NodeInventory{NodeInventory: result},
+		Msg:  &sensor.MsgFromCompliance_NodeInventory{NodeInventory: inventory},
 	}
 	cmetrics.ObserveInventoryProtobufMessage(msg)
-	return msg, nil
+	return msg
+}
+
+// scanNodeWithBackoff runs scans with a linear backoff based on a file to not overstrain a Node if the container keeps restarting.
+// The backoff file will only be encountered if the previous container is killed during a call to scanNodeWithBackoff.
+// Note: This does not prevent strain in case of repeated pod recreation, as it is based on an EmptyDir.
+func scanNodeWithBackoff(opts *cachedScanOpts) (*sensor.MsgFromCompliance, error) {
+	backoffIntervalSeconds := int64(env.NodeInventoryInitialBackoff.DurationSetting() / time.Second)
+
+	backoffFile, err := os.ReadFile(fmt.Sprintf("%s/backoff", opts.inventoryCachePath))
+	defer func() {
+		if err := os.Remove(fmt.Sprintf("%s/backoff", opts.inventoryCachePath)); err != nil {
+			log.Warnf("Could not remove backoff state file: %v", err)
+		}
+	}()
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			log.Debug("No backoff found, continuing without pause")
+		} else {
+			return nil, err
+		}
+	} else {
+		// We have an existing backoff counter
+		backoffIntervalMillis, err := strconv.ParseInt(string(backoffFile), 10, 64)
+		if err != nil {
+			return nil, err
+		}
+		backoffIntervalSeconds = int64(time.Duration(backoffIntervalMillis) / time.Second)
+		log.Infof("Found existing backoff. Waiting before running next inventory")
+		backoffIntervalSeconds = opts.backoffWaitCallback(backoffIntervalSeconds)
+	}
+
+	err = os.WriteFile(fmt.Sprintf("%s/backoff", opts.inventoryCachePath), []byte(fmt.Sprintf("%d", backoffIntervalSeconds)), 0600)
+	if err != nil {
+		return nil, err
+	}
+
+	message, err := cachedScanNode(opts)
+	if err != nil {
+		return nil, err
+	}
+	return message, nil
+}
+
+// cachedScanNode checks for a cached inventory before running a new scan
+func cachedScanNode(opts *cachedScanOpts) (*sensor.MsgFromCompliance, error) {
+	var inventory *storage.NodeInventory
+
+	cachedInv, err := os.ReadFile(fmt.Sprintf("%s/last_scan", opts.inventoryCachePath))
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			log.Debug("No cache file found, running new inventory")
+		}
+		// Running into any other error than ErrNotExist continues to a fresh inventory
+		log.Infof("Unable to read inventory cache, running new inventory. Error: %v", err)
+	} else {
+		// try reading the inventory cached from disk and check whether it is new enough to send
+		cachedInvMsg := &storage.NodeInventory{}
+		if e := proto.Unmarshal(cachedInv, cachedInvMsg); e != nil {
+			// in this case, also collect a fresh inventory
+			log.Infof("Unable to deserialize inventory cache - running new inventory. Error: %v", e)
+		} else {
+			scanTime := cachedInvMsg.GetScanTime()
```

Comment:
How reliable is `ScanTime` attribute on the proto? For caching purposes, should we maybe treat it as something that can go out of our control (scanner repo changes independently) and instead rely on the cache file modification time?

Reply:
I expect `scanTime` to be rather stable, as it is analogous to the `scanTime` field in the `NodeScan` message, which we use extensively in existing code, e.g. to determine whether DB-persisted node information is up to date.
Additionally, while the message definition is in the Scanner repo, the `scanTime` we rely on for cache freshness is set by our own code in the `NodeInventorizer`:

https://github.com/stackrox/stackrox/blob/53f171bbbe7430bd09bccd8fc6c5178d2f04e280/compliance/collection/nodeinventorizer/nodeinventory.go#L59

Reply:
I still find it problematic that value is set in one code file but used in another code file. The reason is that we can move lots of `nodeinventory.go` code into `stackrox/scanner` repo where it will live its own life. If after some time, that code stops providing this value, the cache will stop working correctly.
I suggest a few approaches to deal with that:
1. Overwrite `ScanTime` in the caching logic. Not great as it unnecessarily skews the data.
2. Wrap data returned from `nodeinventory.go` in a serializable struct with its own timestamp field. A bit verbose, but the most reliable.
3. Use file last modification time as a way to detect freshness.

Reply:
I have implemented option 2 and in the process moved to write JSON to disc, which should make debugging easier.
See 2f9a46e8533e176ca2b7e7fa16a6219eb8b12d4c